contracts/cw-shuffle-voting/src/contract.rs

```diff
@@ -0,0 +1,362 @@
+        } => {
+            shuffle_voting_power(deps, env, staked_addresses_count, Some(addr), None)?;
+        }
+        StakeChangedHookMsg::Unstake {
+            addr,
+            amount: _,
+            staked_addresses_count,
+        } => {
+            shuffle_voting_power(deps, env, staked_addresses_count, None, Some(addr))?;
+        }
+    }
+    Ok(())
+}
+
+// Randomly distributes vote power amongst addresses.
+pub fn shuffle_voting_power(
+    deps: DepsMut,
+    env: Env,
+    addresses_count: u64,
+    addr_to_add: Option<Addr>,
+    addr_to_remove: Option<Addr>,
+) -> StdResult<()> {
+    // Generate voting weights based on total staked amount and number of participating addresses.
+    let weights = generate_random_voting_weights(env.block.height, addresses_count);
+
+    // Remove unstaked address if exists
+    if let Some(addr) = addr_to_remove {
+        VOTE_WEIGHTS.remove(deps.storage, &addr, env.block.height)?;
+    }
+
+    // Collect addresses into vector.
+    let mut addresses = VOTE_WEIGHTS
+        .range(deps.storage, None, None, cosmwasm_std::Order::Ascending)
+        .into_iter()
+        .map(|x| match x {
+            Err(_) => Err(x.unwrap_err()),
+            Ok(_) => Ok(x.unwrap().0),
+        })
+        .collect::<StdResult<Vec<Addr>>>()?;
+
+    // Add staked address if exists
+    if let Some(addr) = addr_to_add {
+        addresses.push(addr);
+    }
+
+    // Assign random weights to addresses.
+    for (i, addr) in addresses.into_iter().enumerate() {
+        VOTE_WEIGHTS.save(deps.storage, &addr, &weights[i], env.block.height)?;
+    }
+
+    // Add sum of vote weights to total power map.
+    let sum = VOTE_WEIGHTS
+        .range(deps.storage, None, None, cosmwasm_std::Order::Ascending)
+        .map(|x| match x {
+            Err(_) => Err(x.unwrap_err()),
+            Ok(_) => Ok(x.unwrap().1),
+        })
+        .collect::<StdResult<Vec<Uint128>>>()?
+        .iter()
+        .sum::<Uint128>();
+
+    HEIGHT_TO_TOTAL_POWER.save(deps.storage, env.block.height, &sum)?;
+    Ok(())
+}
+
+// This function uses block height as the seed for a minimal PRNG to generate our random values.
+// This is to allow for every validator node to deterministically derive a sequence of weights.
+// The downside to using block_height, of course, is that our random sequences will be predictable.
+// This may not be a big issue given this doesn't allow for more ways to take over the DAO, just to know in advance the
+// distribution of voting powers. In the future we can explore using an on-chain drand-based beacon as a seed for unpredictable randomness: https://github.com/confio/rand
+pub fn generate_random_voting_weights(block_height: u64, addresses_count: u64) -> Vec<Uint128> {
+    let mut rng = Rand32::new(block_height);
+    let mut random_weights: Vec<Uint128> = vec![];
+    let mut i = 0;
+
+    // Generate addresses_count # of random numbers
+    while i < addresses_count {
```

Comment:
This is going to run out of gas for a large enough number of stakers. Need to work out some way to fix that if we want to deploy this to mainnet. :)